### PR TITLE
Allow users to bridge multiple assets in one transaction

### PR DESCRIPTION
**Pull request by Boyuan-Chen**
* Allow users to bridge multiple assets in one tx
> Gas Usage
> One ER20 Token: 188593
> One ERC20 Token + ETH: 225580 (199984/ (188593 * 2) = 53%)
> Two ERC20 Tokens + ETH: 279921 (249863 / (188593 * 3) = 44%)
* Gateway
image
* Fix the inconsistency between the local deployer and the production

### CODE REVIEW

File: packages/boba/contracts/contracts/LP/L1LiquidityPool.sol [hunk 1]
```diff
@@ -474,6 +488,47 @@ contract L1LiquidityPool is CrossDomainEnabledFast, ReentrancyGuardUpgradeable,
...
+            ClientDepositToken memory token = _tokens[i];
+            PoolInfo storage pool = poolInfo[token.l1TokenAddress];
+            if (token.l1TokenAddress != address(0)) {
+                IERC20(token.l1TokenAddress).safeTransferFrom(msg.sender, address(this), token.amount);
+            } else {
+                ETHAmount = token.amount;
+            }
+            payload[i] = ClientPayToken(msg.sender, pool.l2TokenAddress, token.amount);
+        }
+        require(ETHAmount == msg.value, "Invalid ETH Amount");
```
Comment (InoMurko) [inline packages/boba/contracts/contracts/LP/L1LiquidityPool.sol:511]: this require would only verify the last token amount? is this correct?

Comment (Boyuan-Chen) [reply packages/boba/contracts/contracts/LP/L1LiquidityPool.sol:511]: The last one is to verify that the msg.value is equal to the amount that the user provides in the payload. 
```
  const tx = await Proxy__L1LiquidityPool.clientDepositL1Batch([
    {
      amount: ethers.utils.parseEther('1'),
      l1TokenAddress: ethers.constants.AddressZero
    }
  ], {value: ethers.utils.parseEther('1')})

```

File: integration-tests/test/mrf_lp.spec.ts [hunk 2]
```diff
@@ -1707,9 +1706,606 @@ describe('Liquidity Pool Test', async () => {
...
+      expect(preL2EthBalance).to.deep.eq(postL2EthBalance)
+      expect(postL2ERC20_1Balance).to.deep.eq(
+        preL2ERC20_1Balance.add(
+          depositAmount.mul(remainingERC20_1Percent).div(1000)
+        )
       )
+      expect(preL2ERC20_2Balance).to.deep.eq(postL2ERC20_2Balance)
     })
   })
 })
```
Comment (souradeep-das) [inline integration-tests/test/mrf_lp.spec.ts:2311]: minor: could add another test to verify a batch consisting of all same tokens

File: integration-tests/test/mrf_lp.spec.ts [hunk 3]
```diff
@@ -1707,9 +1706,606 @@ describe('Liquidity Pool Test', async () => {
 
       const postxBOBAAmount = await xBOBAToken.balanceOf(env.l2Wallet.address)
 
-      expect(prexBOBAAmount).to.deep.eq(
-        postxBOBAAmount.add(exitAmount)
+      expect(prexBOBAAmount).to.deep.eq(postxBOBAAmount.add(exitAmount))
+    })
+  })
+
+  describe('Onramp batch tests', async () => {
```
Comment (souradeep-das) [inline integration-tests/test/mrf_lp.spec.ts:1713]: for all the tests could add check to verify the exact amount is deducted from the balance on L1

File: packages/boba/contracts/contracts/LP/L1LiquidityPool.sol [hunk 4]
```diff
@@ -474,6 +488,48 @@ contract L1LiquidityPool is CrossDomainEnabledFast, ReentrancyGuardUpgradeable,
...
+    )
+        external
+        payable
+        whenNotPaused
+    {
+        require(_tokens.length < 5, "Too Many Tokens");
+        uint256 ETHAmount;
+        ClientPayToken[] memory payload = new ClientPayToken[](_tokens.length);
+        for (uint256 i = 0; i < _tokens.length; i++) {
+            ClientDepositToken memory token = _tokens[i];
```
[the remark this answers is not in the document]
Comment (Boyuan-Chen) [reply packages/boba/contracts/contracts/LP/L1LiquidityPool.sol:502]: Add 
```
require(token.amount != 0, "Invalid Amount");
```

File: packages/boba/contracts/contracts/LP/L1LiquidityPool.sol [hunk 5]
```diff
@@ -474,6 +488,48 @@ contract L1LiquidityPool is CrossDomainEnabledFast, ReentrancyGuardUpgradeable,
...
+            PoolInfo storage pool = poolInfo[token.l1TokenAddress];
+            if (token.l1TokenAddress != address(0)) {
+                IERC20(token.l1TokenAddress).safeTransferFrom(msg.sender, address(this), token.amount);
+            } else {
+                ETHAmount = token.amount;
+            }
+            require(pool.l2TokenAddress != address(0), "Invaild Token");
+            payload[i] = ClientPayToken(msg.sender, pool.l2TokenAddress, token.amount);
+        }
+        require(ETHAmount == msg.value, "Invalid ETH Amount");
```
[the remark this answers is not in the document]
Comment (Boyuan-Chen) [reply packages/boba/contracts/contracts/LP/L1LiquidityPool.sol:513]: Updated the code and added tests
```
ETHAmount = ETHAmount + token.amount;
```

File: integration-tests/test/mrf_lp.spec.ts [hunk 6]
```diff
@@ -1707,9 +1706,606 @@ describe('Liquidity Pool Test', async () => {
...
+        env.l2Wallet.address
+      )
+
+      expect(preL2EthBalance).to.deep.eq(postL2EthBalance)
+      expect(postL2ERC20_1Balance).to.deep.eq(
+        preL2ERC20_1Balance.add(
+          depositAmount.mul(remainingERC20_1Percent).div(1000)
+        )
       )
+      expect(preL2ERC20_2Balance).to.deep.eq(postL2ERC20_2Balance)
```
[the remark this answers is not in the document]
Comment (Boyuan-Chen) [reply integration-tests/test/mrf_lp.spec.ts:2472]: Done

File: integration-tests/test/mrf_lp.spec.ts (hunk 2 above)
Comment (Boyuan-Chen) [reply integration-tests/test/mrf_lp.spec.ts:2475]:
```
    it('should deposit ERC20 twice in batch', async () => {
      const depositAmount = utils.parseEther('10')

      const preL1ERC20Balance = await L1ERC20_1.balanceOf(env.l1Wallet.address)
      const preL2ERC20Balance = await L2ERC20_1.balanceOf(env.l2Wallet.address)

      await approveTransaction(
        L1ERC20_1,
        L1LiquidityPool.address,
        depositAmount.mul(BigNumber.from(2))
      )
      const userRewardERC20FeeRate = await L2LiquidityPool.getUserRewardFeeRate(
        L2ERC20_1.address
      )

      const depositTx = await env.waitForXDomainTransaction(
        L1LiquidityPool.clientDepositL1Batch([
          { amount: depositAmount, l1TokenAddress: L1ERC20_1.address },
          { amount: depositAmount, l1TokenAddress: L1ERC20_1.address },
        ]),
        Direction.L1ToL2
      )

      const remainingERC20Percent = await getRemainingPercent(
        userRewardERC20FeeRate
      )

      const postL1ERC20Balance = await L1ERC20_1.balanceOf(env.l1Wallet.address)
      const postL2ERC20Balance = await L2ERC20_1.balanceOf(env.l2Wallet.address)

      expect(preL1ERC20Balance).to.deep.eq(
        postL1ERC20Balance.add(depositAmount.mul(2))
      )
      expect(postL2ERC20Balance).to.deep.eq(
        preL2ERC20Balance.add(
          depositAmount.mul(2).mul(remainingERC20Percent).div(1000)
        )
      )
    })

    it('should deposit ETH twice in batch', async () => {
      const depositAmount = utils.parseEther('10')

      const preL2ETHBalance = await env.l2Wallet.getBalance()
      const userRewardETHFeeRate = await L2LiquidityPool.getUserRewardFeeRate(
        env.ovmEth.address
      )

      const depositTx = await env.waitForXDomainTransaction(
        L1LiquidityPool.clientDepositL1Batch(
          [
            {
              amount: depositAmount,
              l1TokenAddress: ethers.constants.AddressZero,
            },
            {
              amount: depositAmount,
              l1TokenAddress: ethers.constants.AddressZero,
            },
          ],
          { value: depositAmount.mul(2) }
        ),
        Direction.L1ToL2
      )

      const remainingETHPercent = await getRemainingPercent(
        userRewardETHFeeRate
      )
      const postL2ETHBalance = await env.l2Wallet.getBalance()

      expect(postL2ETHBalance).to.deep.eq(
        preL2ETHBalance.add(
          depositAmount.mul(2).mul(remainingETHPercent).div(1000)
        )
      )
    })
```

File: integration-tests/test/mrf_lp.spec.ts (hunk 3 above)
Comment (Boyuan-Chen) [reply integration-tests/test/mrf_lp.spec.ts:1713]: Added

**Review by CAPtheorem — APPROVED** [review]

**Pull request merged**